src/projection.jl

```diff
@@ -31,114 +36,317 @@ function Base.show(io::IO, project::ProjectTo{T}) where {T}
-(::ProjectTo{T})(dx::AbstractZero) where {T} = zero(T)
 
-# Number
-ProjectTo(::T) where {T<:Number} = ProjectTo{T}()
+# Zero
+ProjectTo(::AbstractZero) = ProjectTo{AbstractZero}()
+(::ProjectTo{AbstractZero})(dx) = ZeroTangent()
+
+#####
+##### `Base`
+#####
+
+# Bool
+ProjectTo(::Bool) = ProjectTo{AbstractZero}()
+
+# Numbers
+ProjectTo(::Real) = ProjectTo{Real}()
+ProjectTo(::Complex) = ProjectTo{Complex}()
+ProjectTo(::Number) = ProjectTo{Number}()
+for T in (Float16, Float32, Float64, ComplexF16, ComplexF32, ComplexF64)
+    # Preserve low-precision floats as accidental promotion is a common perforance bug
+    @eval ProjectTo(::$T) = ProjectTo{$T}()
+end
+ProjectTo(x::Integer) = ProjectTo(float(x))
+ProjectTo(x::Complex{<:Integer}) = ProjectTo(float(x))
 (::ProjectTo{T})(dx::Number) where {T<:Number} = convert(T, dx)
 (::ProjectTo{T})(dx::Number) where {T<:Real} = convert(T, real(dx))
 
-# Arrays 
-ProjectTo(xs::T) where {T<:Array} = ProjectTo{T}(; elements=map(ProjectTo, xs))
-function (project::ProjectTo{T})(dx::Array) where {T<:Array}
-    _call(f, x) = f(x)
-    return T(map(_call, project.elements, dx))
+# Arrays{<:Number}
+# If we don't have a more specialized `ProjectTo` rule, we just assume that there is
+# no structure worth preserving. Then any array is acceptable as a gradient.
+function ProjectTo(x::AbstractArray{T}) where {T<:Number}
+    element = ProjectTo(zero(T))
+    # if all our elements are going to zero, then we can short circuit and just send the whole thing
+    element isa ProjectTo{<:AbstractZero} && return element
+    return ProjectTo{AbstractArray}(; element=element, axes=axes(x))
 end
-function (project::ProjectTo{T})(dx::AbstractZero) where {T<:Array}
-    return T(map(proj -> proj(dx), project.elements))
+function (project::ProjectTo{AbstractArray})(dx::AbstractArray{S,M}) where {S,M}
+    T = project_type(project.element)
+    dy = S <: T ? dx : map(project.element, dx)
+    if axes(dy) == project.axes
+        return dy
+    else
+        # The rule here is that we reshape to add or remove trivial dimensions like dx = ones(4,1),
+        # where x = ones(4), but throw an error on dx = ones(1,4) etc.
+        for d in 1:max(M, length(project.axes))
+            size(dy, d) == length(get(project.axes, d, 1)) || throw(_projection_mismatch(project.axes, size(dy)))
+        end
+        return reshape(dy, project.axes)
+    end
 end
-(project::ProjectTo{<:Array})(dx::AbstractArray) = project(collect(dx))
 
-# Arrays{<:Number}: optimized case so we don't need a projector per element
-function ProjectTo(x::T) where {E<:Number,T<:Array{E}}
-    return ProjectTo{T}(; element=ProjectTo(zero(E)), size=size(x))
+# Zero-dimensional arrays -- these have a habit of going missing:
+function (project::ProjectTo{AbstractArray})(dx::Number) # ... so we restore from numbers
+    project.axes isa Tuple{} || sum(length, project.axes) == 1 || throw(_projection_mismatch(project.axes, size(dx)))
+    fill(project.element(dx))
```

Comment:
```suggestion
    return fill(project.element(dx))
```

Is this right for the 
` sum(length, project.axes) == 1` case ?
Which could be  `[x]` or `[x;;]` etc
and this fill is going to push them down to the less nice 0-D array.

I think we want a `reshape` after maybe?
Which seems to work (example in 1..7 for brevity)
```
julia> reshape(fill(1), axes([2;;;]))
1×1×1 Array{Int64, 3}:
[:, :, 1] =
 1
```

Reply:
It has bugs... but my intention was to accept numbers *only* for zero-arrays, not for vectors etc. Because it's zero-arrays which go missing easily, in broadcasting or in getindex. Like adjoint vectors. 
```
julia> ProjectTo(fill(0))(pi)   # this I wanted to work
0-dimensional Array{Float64, 0}:
3.141592653589793

julia> ProjectTo([1])(pi)  # this I want to be an error
0-dimensional Array{Float64, 0}:
3.141592653589793

julia> ProjectTo([1;;;])(pi)  # this too an error, but a better one
ERROR: DimensionMismatch("variable with size(x) == (1, 1, 1) cannot have a gradient with size(dx) == ()")
Stacktrace:
 [1] (::ProjectTo{AbstractArray, NamedTuple{(:element, :axes), Tuple{ProjectTo{Float64, NamedTuple{(), Tuple{}}}, Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}, Base.OneTo{Int64}}}}})(dx::Irrational{:π})
   @ ChainRulesCore ~/.julia/dev/ChainRulesCore/src/projection.jl:160
 [2] top-level scope
   @ REPL[7]:1
```

Reply:
Ok then we can remove the `sum(length, project.axes) == 1` from the condition above?

Reply:
Yea I'll re-check the logic

Reply:
Done, tested